core/chain/abci.go

```diff
@@ -83,19 +85,44 @@ func (app *CoreApplication) CheckTx(_ context.Context, check *abcitypes.CheckTxR
 		return &abcitypes.CheckTxResponse{Code: abcitypes.CodeTypeOK}, nil
 	}
 	// else check if kv store tx, this is hacky and kv store should be in protobuf if we later want to keep it
-	code := app.isValid(check.Tx)
-	return &abcitypes.CheckTxResponse{Code: code}, nil
+	if app.isValidKVTx(check.Tx) {
+		return &abcitypes.CheckTxResponse{Code: abcitypes.CodeTypeOK}, nil
+	} else {
+		return &abcitypes.CheckTxResponse{Code: 1}, nil
+	}
 }
 
 func (app *CoreApplication) InitChain(_ context.Context, chain *abcitypes.InitChainRequest) (*abcitypes.InitChainResponse, error) {
 	return &abcitypes.InitChainResponse{}, nil
 }
 
-func (app *CoreApplication) PrepareProposal(_ context.Context, proposal *abcitypes.PrepareProposalRequest) (*abcitypes.PrepareProposalResponse, error) {
-	return &abcitypes.PrepareProposalResponse{Txs: proposal.Txs}, nil
+func (app *CoreApplication) PrepareProposal(ctx context.Context, proposal *abcitypes.PrepareProposalRequest) (*abcitypes.PrepareProposalResponse, error) {
+	proposalTxs := proposal.Txs
+
+	app.startInProgressTx(ctx)
```

Comment:
We shouldn't do this here nor the commit on line 111, adding it to the list of txs will commit it in the finalize block step. If we commit here then we write to the DB prior to consensus being completed.